examples/llava/llava.cpp

```diff
@@ -2,101 +2,166 @@
 #include "llava-utils.h"
 #include "common.h"
 #include "llama.h"
+#include "llava.h"
 
 #include <cstdio>
 #include <cstdlib>
 #include <vector>
 
+#include "base64.hpp"
+
 static void show_additional_info(int /*argc*/, char ** argv) {
     printf("\n example usage: %s -m <llava-v1.5-7b/ggml-model-q5_k.gguf> --mmproj <llava-v1.5-7b/mmproj-model-f16.gguf> --image <path/to/an/image.jpg> [--temp 0.1] [-p \"describe the image in detail.\"]\n", argv[0]);
     printf("  note: a lower temperature value like 0.1 is recommended for better quality.\n");
 }
 
-int main(int argc, char ** argv) {
-    ggml_time_init();
-
-    gpt_params params;
+static bool encode_image_with_clip(clip_ctx * ctx_clip, int n_threads, const clip_image_u8 * img, float * image_embd, int * n_img_embd, int * n_img_pos, float * t_img_enc_ms) {
+    clip_image_f32 img_res;
+    if (!clip_image_preprocess(ctx_clip, img, &img_res, /*pad2square =*/ true)) {
+        fprintf(stderr, "%s: unable to preprocess image\n", __func__);
 
-    if (!gpt_params_parse(argc, argv, params)) {
-        show_additional_info(argc, argv);
-        return 1;
+        return false;
     }
 
-    if (params.mmproj.empty() || params.image.empty()) {
-        gpt_print_usage(argc, argv, params);
-        show_additional_info(argc, argv);
-        return 1;
+    *n_img_pos = clip_n_patches(ctx_clip);
+    *n_img_embd = clip_n_mmproj_embd(ctx_clip);
+
+    const int64_t t_img_enc_start_us = ggml_time_us();
+    if (!clip_image_encode(ctx_clip, n_threads, &img_res, image_embd)) {
+        fprintf(stderr, "Unable to encode image\n");
+
+        return false;
     }
+    const int64_t t_img_enc_end_us = ggml_time_us();
+    *t_img_enc_ms = (t_img_enc_end_us - t_img_enc_start_us) / 1000.0;
+    return true;
+}
+
+static const char* IMG_BASE64_TAG_BEGIN = "<img src=\"data:image/jpeg;base64,";
+static const char* IMG_BASE64_TAG_END = "\">";
 
-    const char * clip_path = params.mmproj.c_str();
-    const char * img_path = params.image.c_str();
+static void find_image_tag_in_prompt(const std::string& prompt, size_t& begin_out, size_t& end_out) {
```

Comment:
Better to have such functions in `llava-utils.h`

Reply:
👍 done